doc/src/base/sort.md

```diff
@@ -134,88 +138,45 @@ Base.Sort.partialsortperm!
-array of half the size of the input array – and is typically not quite as fast as `QuickSort`.
-It is the default algorithm for non-numeric data.
-
-The default sorting algorithms are chosen on the basis that they are fast and stable.
-Usually, `QuickSort` is selected, but `InsertionSort` is preferred for small data.
-You can also explicitly specify your preferred algorithm, e.g.
-`sort!(v, alg=PartialQuickSort(10:20))`.
-
-The mechanism by which Julia picks default sorting algorithms is implemented via the
-`Base.Sort.defalg` function. It allows a particular algorithm to be registered as the
-default in all sorting functions for specific arrays. For example, here is the default
-method from [`sort.jl`](https://github.com/JuliaLang/julia/blob/master/base/sort.jl):
-
-```julia
-defalg(v::AbstractArray) = DEFAULT_STABLE
-```
-
-You may change the default behavior for specific types by defining new methods for `defalg`.
+You can explicitly specify your preferred algorithm with the `alg` keyword
+(e.g. `sort!(v, alg=PartialQuickSort(10:20))`) or reconfigure the default sorting algorithm
+for a custom types by adding a specialized method to the `Base.Sort.defalg` function.
 For example, [InlineStrings.jl](https://github.com/JuliaStrings/InlineStrings.jl/blob/v1.3.2/src/InlineStrings.jl#L903)
 defines the following method:
 ```julia
 Base.Sort.defalg(::AbstractArray{<:Union{SmallInlineStrings, Missing}}) = InlineStringSort
 ```
 
 !!! compat "Julia 1.9"
-    The default sorting algorithm (returned by `Base.Sort.defalg`) is guaranteed
-    to be stable since Julia 1.9. Previous versions had unstable edge cases when sorting numeric arrays.
+    The default sorting algorithm (returned by `Base.Sort.defalg`) is guaranteed to be stable
+    since Julia 1.9. Previous versions had unstable edge cases when sorting numeric arrays.
 
 ## Alternate orderings
 
-By default, `sort` and related functions use [`isless`](@ref) to compare two
-elements in order to determine which should come first. The
-[`Base.Order.Ordering`](@ref) abstract type provides a mechanism for defining
-alternate orderings on the same set of elements. Instances of `Ordering` define
-a [total order](https://en.wikipedia.org/wiki/Total_order) on a set of elements,
-so that for any elements `a`, `b`, `c` the following hold:
-
-* Exactly one of the following is true: `a` is less than `b`, `b` is less than
-  `a`, or `a` and `b` are equal (according to [`isequal`](@ref)).
-* The relation is transitive - if `a` is less than `b` and `b` is less than `c`
-  then `a` is less than `c`.
+By default, `sort`, `searchsorted`, and related functions use [`isless`](@ref) to compare
+two elements in order to determine which should come first. The
+[`Base.Order.Ordering`](@ref) abstract type provides a mechanism for defining alternate
+orderings on the same set of elements. Instances of `Ordering` define a
+[strict partial order](https://en.wikipedia.org/wiki/Partially_ordered_set#Strict_partial_order).
```

Comment:
As above, strict because we are talking about `<` not `<=`. Partial because that is all that we can actually guarantee. `By(abs)` is an example of a strict partial order that is not a strict total order because none of `lt(By(abs), -1, 1)`, `lt(By(abs), 1, -1)`, and `isequal(-1, 1)` hold.

Weakening the guarantee of totality is a bug fix because we never honored it nor could reasonably have honored it.

Reply:
I think sorting algorithms often require something a bit stronger: a [strict weak ordering](https://en.wikipedia.org/wiki/Weak_ordering). Is it not the case in Julia?

(Compared to a strict partial ordering, a strict weak ordering also requires that `!lt(a,b) && !lt(a,b)` together with `!lt(b,c) && !lt(c,b)` imply `!lt(a,c) && !lt(c,a)`, meaning that the "non-comparability" is transitive.)

Reply:
If `!isless(b,a)` and `!isless(c,b)`, then `sort([a,b,c])` will call `isless` twice and return `[a,b,c]` without checking `isless(c,a)`. So yes, a strict weak ordering is required. Thank you.

Reply:
IIUC it is equally powerful to state that `!lt(a,b) && !lt(b,c)` implies `!lt(a,c)`.

Reply:
> IIUC it is equally powerful to state that `!lt(a,b) && !lt(b,c)` implies `!lt(a,c)`.

I don't think that's right, for example consider the following (invalid) order on complex numbers:

```
lt(a,b) = isreal(a) && isreal(b) ? a<b : abs(a)<abs(b)
```

Then with `a, b, c = 1, im, -1`:

```
julia> !lt(a,b) && !lt(b,c)
true

julia> !lt(a,c)
true
```

but

```
incomparable(a,b) = !lt(a,b) && !lt(b,a)

julia> incomparable(a,b) && incomparable(b,c)
true

julia> incomparable(a,c)
false
```

I think even if we find a way to reduce the conditions to something simpler, it's probably a bad idea: better give the conditions that are standard in the literature and that make the point clear, for example here the point is "transitiveness of incomparability", which would not be obvious if we gave a kind of reduced condition...

Reply:
That is not a valid counterexample because for `a, b, c = 0, -1, im` we have
```
julia> !lt(a,b) && !lt(b,c)
true

julia> !lt(a,c)
true
```

To prove that _transitivity of not less than_ implies _transitivity of incomparability_, take arbitrary `lt` which satisfies `!lt(a,b) && !lt(b,c) implies !lt(a,c)`. Take arbitrary `a`, `b`, `c` with `incomparable(a, b) && incomparable(b, c)`, expand to `(!lt(a,b) && !lt(b,a)) && (!lt(b,c) && !lt(c,b))`, re-arrange to `(!lt(a,b) && !lt(b,c)) && (!lt(c,b) && !lt(b,a))`, and apply transitivity of not less than to each half to get `!lt(a,c) && !lt(c,a)` which is `incomparable(a,c)`. Because `a`, `b`, and `c` were arbitrary that satisfy `incomparable(a, b) && incomparable(b, c)`, we know `lt` satisfies _transitivity of incomparability_. Because `lt` was and arbitrary function with `!lt(a,b) && !lt(b,c) implies !lt(a,c)`, we have that _transitivity of not less than_ implies _transitivity of incomparability_.

The converse is more difficult. Take arbitrary `lt` which satisfies _transitivity_ and _transitivity of incomparability_. Take arbitrary `a`, `b`, `c` with `!lt(a, b) && !lt(b, c)`. Assume to reach a contradiction that `lt(a,c)`. By antisymetry, we have `!lt(c,a)`. If `lt(b, a)`, then we could apply transitivity to claim `lt(b,c)` which is false, and similarly if `lt(c,b)` then we could apply transitivity to claim `lt(c,a)` which is also false. Consequuently, we have `!lt(b,a)` and `!lt(c,b)`. We can now apply _transitivity of incomaprability_ to claim `a` and `c` are incomparable which contradicts `lt(a,c)`. Therefore our original assumption was false so `!lt(a,c)`. Because `a`, `b`, and `c` were arbitrary with `!lt(a, b) && !lt(b, c)`, we know that `lt` satisfies _transitivity of not less than_. Because `lt` was arbitrary with _transitivity_ and _transitivity of incomparability_, we know that _transitivity_ together with _transitivity of incomparability_ implies _transitivity of not less than_.

In conclusion, if we already have transitivity (which we do) "`!lt(a,b)` && `!lt(b,c)` implies `!lt(a,c)`" and "`incomparable(a,b) && incomarable(b,c)` implies `incomarable(a,c)`" are mathematically interchangeable.

It is equally valid to think of the criterion as "transitivity of incomparability" and "transitivity of not less than", though switching back and forth rigorously is nontrivial.

That said, I agree with you that incomparability is a better criterion to pick for intuition.

Reply:
Thanks for the correction, my counterexample was wrong indeed. And I understand now that you were talking of equivalence in the context where the other conditions (in practice, transitivity) are satisfied.